Readme.md

# ext.js - JavaScript Extensions &amp; Utilities

High quality JavaScript extensions for various tasks such as
formatting dates, currency helpers, and more. This library targets node.js and
server-side frameworks such as [Express](http://github.com/visionmedia/express).

Ext.js uses the CommonJS module pattern when needed.
Additionally, **all** of the native object extensions are _non-enumerable_,
and you may create your own via `require('ext').extend(prototype, methods).

## Installation

  Install the [Kiwi package manager for nodejs](http://github.com/visionmedia/kiwi)
  and run:
  
      $ kiwi install ext

## Extensions

To use simply:
    require('ext')

### Object

  * Object#keys
  * Object#values
  * Object#join([str])
  * Object#alias(orig, alias)
  * Object#tap(fn[, context])
  * Object#merge(obj)
  * Object#mergeDeep(obj)
  * Object#each(fn[, context])
  * Object#respondsTo(key)

### Array

  * Array#each(fn[, context])
  * Array#excludes(item[, item, ...])
  * Array#includes(item[, item, ...])
  * Array#clear
  * Array#flattened
  * Array#first
  * Array#last
  * Array#sample
  * Array#compact([removableValues])
  * Array#at(index)
  * Array#drop(n)
  * Array#take(n)
  * Array#grep(pattern)
  * Array#remove(obj[, fn[, context]])
  * Array#transposed
  * Array#isEmpty

#### Collection Operators

  * Array#sum
  * Array#avg
  * Array#min
  * Array#max

### Enumerable (included by Object and Array)

  * Enumerable#map(fn[, context])
  * Enumerable#reduce(fn[, memo]) aliased as inject()
  * Enumerable#filter(fn[, context]) aliased as select()
  * Enumerable#detect(fn[, context]) aliased as find()
  * Enumerable#some(fn[, context]) aliased as any()
  * Enumerable#every(fn[, context]) aliased as all()
  * Enumerable#none(fn[, context])
  * Enumerable#reject(fn[, context])

### Number

  * Number#ordinalize
  * Number#currency
  * Number#second     / Number#seconds
  * Number#minute     / Number#minutes
  * Number#hour       / Number#hours
  * Number#day        / Number#days
  * Number#week       / Number#weeks
  * Number#month      / Number#months
  * Number#year       / Number#years
  * Number#toSeconds
  * Number#toMinutes
  * Number#toHours
  * Number#toDays
  * Number#toWeeks
  * Number#toMonths
  * Number#toYears
  * Number#byte       / Number#bytes
  * Number#kilobyte   / Number#kilobytes
  * Number#megabyte   / Number#megabytes
  * Number#gigabyte   / Number#gigabytes
  * Number#ago
  * Number#times(fn[, context])
  * Number#isFloat
  * Number#hex
  * Number#octal

### String

  * String#variable
  * String#uppercase
  * String#lowercase
  * String#camelcase
  * String#digits
  * String#strip
  * String#drop(n)
  * String#take(n)
  * String#before(str)
  * String#after(str)
  * String#padLeft(width[, char])
  * String#padRight(width[, char])
  * String#remove(pattern)
  * String#startsWith(str)
  * String#endsWith(str)
  * String#capitalize([all])
  * String#wrap(prefix[, suffix])
  * String#singular
  * String#plural
  * String#isPlural
  * String#isSingular
  * String#includes(str)
  * String#count(str)

### Date

  * Date#year
  * Date#month
  * Date#date
  * Date#day
  * Date#hours
  * Date#minutes
  * Date#seconds
  * Date#milliseconds
  * Date#monthName
  * Date#shortMonthName
  * Date#dayName
  * Date#shortDayName
  * Date#format(str)
  * Date#inWordsSince(date)
  * Date#inWordsSinceNow
  * Date#parse(str)
  * parse(str[, date])

### RegExp

  * RegExp.escape(str[, chars])

### Function

  * Function#bind(context[, ...])
  * Function#curry(...)

### Base64

  * String#base64Encoded / encode(str)
  * String#base64Decoded / decode(str)

### Error

  * Error.raise([name[, message[, object]]])

## MD5

  * String#md5 / hash(str)

### printf

  * sprintf(str[, arg[, ...]])
  * eprintf(str[, arg[, ...]])
  * printf(str[, arg[, ...]])

## Iterator Functions

Ext.js allows most iterators to use _"string lambda expressions"_,
which allows you to define a function by passing a string containing,
a valid Lambda expression or shortform as shown in some of the examples
below.

### Full Lambda Expressions

    [1,2,3].map('a -> a + 1')
    // => [2,3,4]

### Full Lambda Expressions With Currying

    curried = [1,2,3].map('a -> b -> a + b')
    // => [(Function), (Function), (Function)]

    curried[1](40)
    // => 42

### Literals

    [1,2,3,4,5].select(function(n){ return n > 3 })
    // => [4,5]

### Properties

    ['hello', 'there'].map('.length')
    // => [5,5]

    [['hello'], ['there']].map('[0]')
    // => ['hello', 'there']

### Methods

    [1,2,3].map('.toString()')
    // => ['1','2','3']

### Binary Operators

    [1,2,3,4,5].select('> 3')
    // => [4,5]

### Object Usage

    { foo: 'bar', baz: 'something' }.select(".length > 3")
    // => { baz: 'something' }

## Date Parsing

The module _ext/date_ exports the **parse()** function which accepts
a _string_ to parse, as well as an optional _date_ which represents "now".

Below are some examples:

    new Date
    // => Fri, 26 Feb 2010 19:16:47 GMT

    parse('today')
    // => Fri, 26 Feb 2010 19:16:47 GMT

    parse('yesterday')
    // => Thu, 25 Feb 2010 19:16:47 GMT

    parse('in 5 hours')
    // => Sat, 27 Feb 2010 00:16:47 GMT

    parse('in 2 days')
    // => Sun, 28 Feb 2010 19:16:47 GMT

    parse('next tuesday')
    // => Tue, 02 Mar 2010 19:16:47 GMT

More examples:

    'in a year'
    'in one year'
    'in five days'
    'in three hundred minutes'
    'in fifty two minutes'
    'in seventy five trillion seconds'
    'in five and a half minutes'

## Running Tests

    $ make test

## Contributors

  * TJ Holowaychuk (visionmedia)
  * Tobias Svensson (tobiassvn)
  
### Style Guide

These guidlines must be met before commit(s) or patches 
will be accepted.

  * Use 2 space indents
  * No trailing whitespace
  * Blank line before EOF
  * Omit semi-colons unless required (very rarely is this needed)
  * Never bump the version
  
File comments should take the following form:

    // ext.js - <category> [- subcategory ...] - Copyright TJ Holowaychuk <tj@vision-media.ca> (MIT Licensed)

Method comments should take the following form, and
<type>'s should be lowercase. _"mixed"_ should be used
when multiple types are acceptable, **not** _"object"_.

When accepting an _options_ object for example, you
may use _"hash"_ as this indicates that common usage
of this parameter is to pass a literal object using
brace syntax.

All examples should be indented 2 spaces.

    /**
     * <description>
     *
     *  <example ...>
     *
     * @param  {<type>} <name>
     * @param  {<type>} <name>
     * @param  {<type>} <name>
     * @return {<type>}
     * @api <public|private|protected>
     */

No trailing whitespace after hash keys:

    { foo: 'bar' }
    // good
    
    { foo : 'bar' }
    // bad
    
Hash whitespace:

    { foo: 'bar', baz: 'raz' }
    // good
    
    {foo:'bar', baz: 'raz'}
    // bad
    
    {
      foo: 'bar',
      baz: 'raz'
    }
    // good
    
Chained methods should be indented to indicate context:

    str
      .strip
      .replace(...)
      .replace(...)
    // good
    
    str
    .strip
    .replace(...)
    .replace(...)
    // bad
    
    str.
      strip.
      replace(...).
      replace(...)
    // bad
    
Use single quoted strings when possible:

    'yay'
    // good
    
    "he said 'yay'"
    // good
    
    "yay"
    // bad

Large ternary conditional should take the following form:

    foo = some.largeProperty === undefined
      ? 'some large value'
      : some.largeProperty + 'whatever'
    
Conditionals should be functional when possible (I hate braces):

    if (foo)
      bar(),
      baz()
    else
      if (somethingElse)
        whateves()
    // goood
    
    if (foo) {
      bar()
      baz()
    }
    else {
      if (somethingElse) {
        whateves()
      }
    }
    // bad
    
Closures:

    function(){
      
    }
    // good
    
    function () {
      
    }
    // bad
    
Methods:

    foo.bar = function() {
      
    }
    // good
    
    foo.bar = function (){
      
    }
    // bad 
  

## License

(The MIT License)

Copyright (c) 2009 TJ Holowaychuk &lt;tj@vision-media.ca&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



History.md

0.5.0 / 2010-04-23
==================

  * Added Number#octal
  * Added Number#hex
  * Increase performance Object#each() by %50
  * Increase performance of Object#merge()

0.4.1 / 2010-04-20
==================

  * Object#reject() and Object#filter() now return arrays at all times

0.4.0 / 2010-04-20
==================

  * Changed; renamed String#base64Encode -> String#base64Encoded
  * Removed String lambdas
  * Removed String#digitize
  * Removed Array#flatten

0.3.0 / 2010-04-14
==================

  * Added Object#mergeDeep()
  * Fixed issue with Object#{keys,values} having no setters once assigned a value
  * Changed; String#strip now using native String#trim()
  * Changed; Object#merge() accepts undefined, defaulting to {}
  * Changed; Object#mergeDeep() accepts undefined, defaulting to {}

0.2.6 / 2010-03-18
==================

  * Added Number#{gigabyte,megabyte,kilobyte,byte}[s]
  * Changed; String#lambda etc no longer depricated. Benchmarks proved to be invalid
  * Depricated String#digitize in favour of String#digits
  * Depricated Array#flatten in favour of Array#flattened

0.2.5 / 2010-03-16
==================

  * Removed Object#include()
  * Removed last process.mixin() call
  * Deprecated String#lambda() etc. Closes #64

0.2.4 / 2010-03-15
==================

  * APIs including Number#{seconds,minutes,hours, ...} working in milliseconds not seconds. Closes #63

0.2.3 / 2010-03-15
==================

  * Added Function#curry(), fixed Enumerable#map() which always returns an array now
  * Removed some process.mixin() calls
  * Fixed Object#{keys,values}; can be properly overwritten

0.2.2 / 2010-03-11
==================

  * Fixed Object#keys= allowing custom values to take precedence over Object#keys

0.2.1 / 2010-03-10
==================

  * Added Object#join([str])
  * Fixed reduce() bug in which Object#reduce() was previously skipping
    the first value when a memo object was passed.

0.2.0 / 2010-03-04
==================

  * Added Object#alias()
  * Added Object#include().
  * Added Object#reduce()
  * Aliased Enumerable#reduce() as Enumerable#inject()
  * Added Array#transposed
  * Added String#apply()
  * Added String#call()
  * Added String#lambda
  * Added seed.yml for kiwi support

0.1.0 / 2010-03-01
==================

  * Pretty much everything :)

0.0.1 / 2010-01-05
==================

  * Initial release



lib/ext/md5.js

// ext.js - md5 - Copyright TJ Holowaychuk <tj@vision-media.ca> (MIT Licensed)

/**
 *  MD5 (Message-Digest Algorithm)
 *  http://www.webtoolkit.info/
 *
 * @param  {string} string
 * @return {string}
 * @see String#md5
 * @api public
 */

exports.hash = function(string) {

    function RotateLeft(lValue, iShiftBits) {
        return (lValue<<iShiftBits) | (lValue>>>(32-iShiftBits));
    }

    function AddUnsigned(lX,lY) {
        var lX4,lY4,lX8,lY8,lResult;
        lX8 = (lX & 0x80000000);
        lY8 = (lY & 0x80000000);
        lX4 = (lX & 0x40000000);
        lY4 = (lY & 0x40000000);
        lResult = (lX & 0x3FFFFFFF)+(lY & 0x3FFFFFFF);
        if (lX4 & lY4) {
            return (lResult ^ 0x80000000 ^ lX8 ^ lY8);
        }
        if (lX4 | lY4) {
            if (lResult & 0x40000000) {
                return (lResult ^ 0xC0000000 ^ lX8 ^ lY8);
            } else {
                return (lResult ^ 0x40000000 ^ lX8 ^ lY8);
            }
        } else {
            return (lResult ^ lX8 ^ lY8);
        }
    }

    function F(x,y,z) { return (x & y) | ((~x) & z); }
    function G(x,y,z) { return (x & z) | (y & (~z)); }
    function H(x,y,z) { return (x ^ y ^ z); }
    function I(x,y,z) { return (y ^ (x | (~z))); }

    function FF(a,b,c,d,x,s,ac) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(F(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    };

    function GG(a,b,c,d,x,s,ac) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(G(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    };

    function HH(a,b,c,d,x,s,ac) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(H(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    };

    function II(a,b,c,d,x,s,ac) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(I(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    };

    function ConvertToWordArray(string) {
        var lWordCount;
        var lMessageLength = string.length;
        var lNumberOfWords_temp1=lMessageLength + 8;
        var lNumberOfWords_temp2=(lNumberOfWords_temp1-(lNumberOfWords_temp1 % 64))/64;
        var lNumberOfWords = (lNumberOfWords_temp2+1)*16;
        var lWordArray=Array(lNumberOfWords-1);
        var lBytePosition = 0;
        var lByteCount = 0;
        while ( lByteCount < lMessageLength ) {
            lWordCount = (lByteCount-(lByteCount % 4))/4;
            lBytePosition = (lByteCount % 4)*8;
            lWordArray[lWordCount] = (lWordArray[lWordCount] | (string.charCodeAt(lByteCount)<<lBytePosition));
            lByteCount++;
        }
        lWordCount = (lByteCount-(lByteCount % 4))/4;
        lBytePosition = (lByteCount % 4)*8;
        lWordArray[lWordCount] = lWordArray[lWordCount] | (0x80<<lBytePosition);
        lWordArray[lNumberOfWords-2] = lMessageLength<<3;
        lWordArray[lNumberOfWords-1] = lMessageLength>>>29;
        return lWordArray;
    };

    function WordToHex(lValue) {
        var WordToHexValue="",WordToHexValue_temp="",lByte,lCount;
        for (lCount = 0;lCount<=3;lCount++) {
            lByte = (lValue>>>(lCount*8)) & 255;
            WordToHexValue_temp = "0" + lByte.toString(16);
            WordToHexValue = WordToHexValue + WordToHexValue_temp.substr(WordToHexValue_temp.length-2,2);
        }
        return WordToHexValue;
    };

    function Utf8Encode(string) {
        string = string.replace(/\r\n/g,"\n");
        var utftext = "";

        for (var n = 0; n < string.length; n++) {

            var c = string.charCodeAt(n);

            if (c < 128) {
                utftext += String.fromCharCode(c);
            }
            else if((c > 127) && (c < 2048)) {
                utftext += String.fromCharCode((c >> 6) | 192);
                utftext += String.fromCharCode((c & 63) | 128);
            }
            else {
                utftext += String.fromCharCode((c >> 12) | 224);
                utftext += String.fromCharCode(((c >> 6) & 63) | 128);
                utftext += String.fromCharCode((c & 63) | 128);
            }

        }

        return utftext;
    };

    var x=Array();
    var k,AA,BB,CC,DD,a,b,c,d;
    var S11=7, S12=12, S13=17, S14=22;
    var S21=5, S22=9 , S23=14, S24=20;
    var S31=4, S32=11, S33=16, S34=23;
    var S41=6, S42=10, S43=15, S44=21;

    string = Utf8Encode(string);

    x = ConvertToWordArray(string);

    a = 0x67452301; b = 0xEFCDAB89; c = 0x98BADCFE; d = 0x10325476;

    for (k=0;k<x.length;k+=16) {
        AA=a; BB=b; CC=c; DD=d;
        a=FF(a,b,c,d,x[k+0], S11,0xD76AA478);
        d=FF(d,a,b,c,x[k+1], S12,0xE8C7B756);
        c=FF(c,d,a,b,x[k+2], S13,0x242070DB);
        b=FF(b,c,d,a,x[k+3], S14,0xC1BDCEEE);
        a=FF(a,b,c,d,x[k+4], S11,0xF57C0FAF);
        d=FF(d,a,b,c,x[k+5], S12,0x4787C62A);
        c=FF(c,d,a,b,x[k+6], S13,0xA8304613);
        b=FF(b,c,d,a,x[k+7], S14,0xFD469501);
        a=FF(a,b,c,d,x[k+8], S11,0x698098D8);
        d=FF(d,a,b,c,x[k+9], S12,0x8B44F7AF);
        c=FF(c,d,a,b,x[k+10],S13,0xFFFF5BB1);
        b=FF(b,c,d,a,x[k+11],S14,0x895CD7BE);
        a=FF(a,b,c,d,x[k+12],S11,0x6B901122);
        d=FF(d,a,b,c,x[k+13],S12,0xFD987193);
        c=FF(c,d,a,b,x[k+14],S13,0xA679438E);
        b=FF(b,c,d,a,x[k+15],S14,0x49B40821);
        a=GG(a,b,c,d,x[k+1], S21,0xF61E2562);
        d=GG(d,a,b,c,x[k+6], S22,0xC040B340);
        c=GG(c,d,a,b,x[k+11],S23,0x265E5A51);
        b=GG(b,c,d,a,x[k+0], S24,0xE9B6C7AA);
        a=GG(a,b,c,d,x[k+5], S21,0xD62F105D);
        d=GG(d,a,b,c,x[k+10],S22,0x2441453);
        c=GG(c,d,a,b,x[k+15],S23,0xD8A1E681);
        b=GG(b,c,d,a,x[k+4], S24,0xE7D3FBC8);
        a=GG(a,b,c,d,x[k+9], S21,0x21E1CDE6);
        d=GG(d,a,b,c,x[k+14],S22,0xC33707D6);
        c=GG(c,d,a,b,x[k+3], S23,0xF4D50D87);
        b=GG(b,c,d,a,x[k+8], S24,0x455A14ED);
        a=GG(a,b,c,d,x[k+13],S21,0xA9E3E905);
        d=GG(d,a,b,c,x[k+2], S22,0xFCEFA3F8);
        c=GG(c,d,a,b,x[k+7], S23,0x676F02D9);
        b=GG(b,c,d,a,x[k+12],S24,0x8D2A4C8A);
        a=HH(a,b,c,d,x[k+5], S31,0xFFFA3942);
        d=HH(d,a,b,c,x[k+8], S32,0x8771F681);
        c=HH(c,d,a,b,x[k+11],S33,0x6D9D6122);
        b=HH(b,c,d,a,x[k+14],S34,0xFDE5380C);
        a=HH(a,b,c,d,x[k+1], S31,0xA4BEEA44);
        d=HH(d,a,b,c,x[k+4], S32,0x4BDECFA9);
        c=HH(c,d,a,b,x[k+7], S33,0xF6BB4B60);
        b=HH(b,c,d,a,x[k+10],S34,0xBEBFBC70);
        a=HH(a,b,c,d,x[k+13],S31,0x289B7EC6);
        d=HH(d,a,b,c,x[k+0], S32,0xEAA127FA);
        c=HH(c,d,a,b,x[k+3], S33,0xD4EF3085);
        b=HH(b,c,d,a,x[k+6], S34,0x4881D05);
        a=HH(a,b,c,d,x[k+9], S31,0xD9D4D039);
        d=HH(d,a,b,c,x[k+12],S32,0xE6DB99E5);
        c=HH(c,d,a,b,x[k+15],S33,0x1FA27CF8);
        b=HH(b,c,d,a,x[k+2], S34,0xC4AC5665);
        a=II(a,b,c,d,x[k+0], S41,0xF4292244);
        d=II(d,a,b,c,x[k+7], S42,0x432AFF97);
        c=II(c,d,a,b,x[k+14],S43,0xAB9423A7);
        b=II(b,c,d,a,x[k+5], S44,0xFC93A039);
        a=II(a,b,c,d,x[k+12],S41,0x655B59C3);
        d=II(d,a,b,c,x[k+3], S42,0x8F0CCC92);
        c=II(c,d,a,b,x[k+10],S43,0xFFEFF47D);
        b=II(b,c,d,a,x[k+1], S44,0x85845DD1);
        a=II(a,b,c,d,x[k+8], S41,0x6FA87E4F);
        d=II(d,a,b,c,x[k+15],S42,0xFE2CE6E0);
        c=II(c,d,a,b,x[k+6], S43,0xA3014314);
        b=II(b,c,d,a,x[k+13],S44,0x4E0811A1);
        a=II(a,b,c,d,x[k+4], S41,0xF7537E82);
        d=II(d,a,b,c,x[k+11],S42,0xBD3AF235);
        c=II(c,d,a,b,x[k+2], S43,0x2AD7D2BB);
        b=II(b,c,d,a,x[k+9], S44,0xEB86D391);
        a=AddUnsigned(a,AA);
        b=AddUnsigned(b,BB);
        c=AddUnsigned(c,CC);
        d=AddUnsigned(d,DD);
    }

    var temp = WordToHex(a)+WordToHex(b)+WordToHex(c)+WordToHex(d);

    return temp.toLowerCase();
}

// --- Extensions

Ext.extend(String.prototype, {

  /**
   * @see hash()
   */

  get md5() { return exports.hash(this) }
})
